http/backup_service.go

```diff
@@ -0,0 +1,223 @@
+		return
+	}
+
+	internalBackupPath := h.BackupService.InternalBackupPath(id)
+	metaFilename := fmt.Sprintf("%s/influxd.bolt", internalBackupPath)
+	metaFile, err := os.OpenFile(metaFilename, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0660)
+	if err != nil {
+		err = multierr.Append(err, os.RemoveAll(internalBackupPath))
+		h.HandleHTTPError(ctx, err, w)
+		return
+	}
+
+	err = h.KVBackupService.Backup(ctx, metaFile)
+	if err != nil {
+		err = multierr.Append(err, os.RemoveAll(internalBackupPath))
+		h.HandleHTTPError(ctx, err, w)
+		return
+	}
+	files = append(files, "influxd.bolt")
+
+	b := backup{
+		ID:    id,
+		Files: files,
+	}
+	err = json.NewEncoder(w).Encode(&b)
+	if err != nil {
+		err = multierr.Append(err, os.RemoveAll(internalBackupPath))
+		h.HandleHTTPError(ctx, err, w)
+		return
+	}
+}
+
+func (h *BackupHandler) handleFetchFile(w http.ResponseWriter, r *http.Request) {
+	span, r := tracing.ExtractFromHTTPRequest(r, "BackupHandler.handleFetchFile")
+	defer span.Finish()
+
+	ctx := r.Context()
+	defer r.Body.Close()
+
+	params := httprouter.ParamsFromContext(ctx)
+	backupID, err := strconv.Atoi(params.ByName("backup_id"))
+	if err != nil {
+		h.HandleHTTPError(ctx, err, w)
+		return
+	}
+	backupFile := params.ByName("backup_file")
+
+	err = h.BackupService.FetchBackupFile(ctx, backupID, backupFile, w)
+	if err != nil {
+		h.HandleHTTPError(ctx, err, w)
+		return
+	}
+}
+
+type BackupService struct {
+	Addr               string
+	Token              string
+	InsecureSkipVerify bool
+}
+
+func (s *BackupService) CreateBackup(ctx context.Context) (int, []string, error) {
+	span, ctx := tracing.StartSpanFromContext(ctx)
+	defer span.Finish()
+
+	u, err := NewURL(s.Addr, backupPath)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, u.String(), nil)
+	if err != nil {
+		return 0, nil, err
+	}
+	SetToken(s.Token, req)
+	req = req.WithContext(ctx)
+
+	hc := NewClient(u.Scheme, s.InsecureSkipVerify)
+	resp, err := hc.Do(req)
+	if err != nil {
+		return 0, nil, err
+	}
+	defer resp.Body.Close()
+
+	if err := CheckError(resp); err != nil {
+		return 0, nil, err
+	}
+
+	var b backup
+	err = json.NewDecoder(resp.Body).Decode(&b)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	return b.ID, b.Files, nil
+}
+
+func (s *BackupService) FetchBackupFile(ctx context.Context, backupID int, backupFile string, w io.Writer) error {
+	span, ctx := tracing.StartSpanFromContext(ctx)
+	defer span.Finish()
+
+	u, err := NewURL(s.Addr, composeBackupFilePath(backupID, backupFile))
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+	if err != nil {
+		return err
+	}
+	SetToken(s.Token, req)
+
+	hc := NewClient(u.Scheme, s.InsecureSkipVerify)
+	resp, err := hc.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if err := CheckError(resp); err != nil {
+		return err
+	}
+
+	_, err = io.CopyBuffer(w, resp.Body, make([]byte, 1024*1024))
```

Comment:
Did you test that this was any better than just wrapping `w` in a buffered writer (`bufio.Writer`)? It probably is; I just wondered. I would probably make it a constant.

Reply:
I did some reading.

Turns out that `io.CopyBuffer` doesn't use the `buf` parameter when the writer implements `io.WriterTo` or the reader implements `io.ReaderFrom`. `os.File` implements both. In this case (API client), the writer is always a file. On the service side, also in this PR, the reader is always a file. Therefore, I've replaced both instances with `io.Copy`, which ultimately does the same thing.

https://github.com/golang/go/issues/32276